src/Chapter3.hs

```diff
@@ -479,6 +523,36 @@ After defining the city, implement the following functions:
    complicated task, walls can be built only if the city has a castle
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
+type People = Int
+
+data Castle = Castle String | NoCastle deriving (Show)
+data Wall = Wall | NoWall deriving (Show)
+data SpecBuilding = Church | Library deriving (Show)
+data House = House People deriving (Show)
+
+data City = City
+  { cityCastle          :: Castle
+  , cityWall            :: Wall
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

Walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

This nicely translates to the sum type, so you can think of the way to extend yours to show the missing state 🙂 🏰 🧱